File: lab3/python/dh_user.py
from galois_field import GaloisField

from random import randint
import sys

from typing import Generic, TypeVar, List

T = TypeVar("T", bound=GaloisField)

class DHUser(Generic[T]):
    def __init__(self, dhsetup):
        self.secret = randint(0, sys.maxsize)
        self.dhsetup = dhsetup
        self.key = None

    def getPublicKey(self) -> T:
        return self.dhsetup.power(self.dhsetup.getGenerator(), self.secret)

    def setPublicKey(self, key: T):
        self.key = self.dhsetup.power(key, self.secret)

    def encrypt(self, message: T) -> T:
        if self.key == None:
            raise Exception("Cannot encrypt a message without a private key")
        return self.key * message

    def decrypt(self, message: T) -> T:
        if self.key == None:
            raise Exception("Cannot decrypt a message without a private key")
        return message / self.key

File: lab3/java/src/main/java/com/diffiehellman/DHUser.java
package com.diffiehellman;

import java.util.Random;

import com.diffiehellman.GaloisField;
import com.diffiehellman.DHSetup;

public class DHUser<T extends GaloisField> {
    private static Random random = new Random();

    private DHSetup<T> setup;
    private long secret;
    private T key;

    public DHUser(DHSetup<T> setup) {
        this.setup = setup;
        this.secret = Math.abs(random.nextLong());
        this.key = null;
    }

    public void setPublicKey(T key) {
        this.key = this.setup.power(key, secret);
    }

    public T getPublicKey() {
        return this.setup.power(this.setup.getGenerator(), this.secret);
    }

    public T encrypt(final T message) throws IllegalStateException {
        if (this.key == null)
            throw new IllegalStateException("Cannot encrypt a message without a private key");

        return (T) message.multiply(key);
    }

    public T decrypt(final T message) throws IllegalStateException {
        if (this.key == null)
            throw new IllegalStateException("Cannot decrypt a code without a private key");

        return (T) message.divide(key);
    }
}
